frontend/src/app/auth/auth.service.ts

```diff
@@ -52,16 +54,49 @@ export class AuthService {
         if (localStorage.removeItem(this.localStorageKey)) {
             // todo: place logout request to backend
         }
+        this.router.navigate(['login']);
     }
 
-    getToken(): string {
-        return JSON.parse(localStorage.getItem(this.localStorageKey)).access_token;
+    getToken() {
+        const currentTime = (new Date()).getTime();
+        const timeoutTime = (+localStorage.getItem(this.localStorageKeyRefreshTimeout));
+        if (timeoutTime < currentTime) {
+            return this.refreshToken().map(response => response.access_token);
+        }
+        return Observable.of(JSON.parse(localStorage.getItem(this.localStorageKey)).access_token);
     }
 
     isLoggedIn(): boolean {
         return !!localStorage.getItem(this.localStorageKey);
+    }
 
+    refreshToken() {
+        if (!this.isRefreshingToken) {
+            this.isRefreshingToken = true;
+            const httpOptions = {
+                headers: new HttpHeaders({
+                    'Authorization': 'Basic ' + btoa(this.client_id + ':' + this.secret),
+                    'Content-Type': 'application/x-www-form-urlencoded'
+                })
+            };
+            const dataString = 'grant_type=refresh_token&client_id=' + this.client_id + '&refresh_token='
+                + JSON.parse(localStorage.getItem(this.localStorageKey)).refresh_token;
+            return this.http.post<Oauth>(this.tokenURL, dataString, httpOptions)
+                .do(response => {
+                    if (response && response.access_token) {
+                        localStorage.setItem(this.localStorageKey, JSON.stringify(response));
+                        const timeout = '' + (((response.expires_in) * 1000) + (new Date()).getTime());
+                        localStorage.setItem(this.localStorageKeyRefreshTimeout, timeout);
+                        this.isRefreshingToken = false;
+                    } else {
+                        this.isRefreshingToken = false;
```

Comment:
Same in both branches, could be moved before the if statement to reduce duplication

Reply:
Nevermind